Cython 3 updates and test build linker flags

karlotness opened the pull request:
It looks like omitting the linker flags when during function detection causes detection to fail for statically linked pthread builds. Passing them through seems to fix the issues.

This should be related to #294 (and also conda-forge/pyfftw-feedstock#51 and conda-forge/pyfftw-feedstock#59).

Thanks!

karlotness commented:
I've only been able to take brief look. It seems like it may have been left commented out inadvertently. That change was made as part of bfde139 which seems to be focused on changing how stdout/stderr are captured, and I didn't see a mention of changes to the flags argument in the associated PR.

hgomersall commented:
Let's see what happens in the CI pipeline...

hgomersall commented:
Hmmm, it's not looking good. Are you able to debug this? It's failing at a build step and I'm currently rather snowed under.

karlotness commented:
Those build failures look like issues with some of the changes in Cython 3. Probably the easiest thing to do in the short term is something like #359 to stay with the older version until it's possible to upgrade. Those two changes together are effectively conda-forge/pyfftw-feedstock#60 which does at least seem to go through ok.

hgomersall commented:
> Those build failures look like issues with some of the changes in Cython 3. Probably the easiest thing to do in the short term is something like #359 to stay with the older version until it's possible to upgrade. Those two changes together are effectively [conda-forge/pyfftw-feedstock#60](https://github.com/conda-forge/pyfftw-feedstock/pull/60) which does at least seem to go through ok.

Sounds good, do you fancy making the changes?

karlotness commented:
I've added some adjustments on top to (I think) fix the issues with Cython 3. The diff certainly looks bigger than it is, a big chunk of it is just reducing indentation.

The main thing was adding some `noexcept`/`nogil` annotations since that was what caused most of the actual errors.

There were also a good number of deprecation warnings about the conditional compilation (`IF`s) since those have been deprecated in Cython 3 (see cython/cython#4310). To move away from those:
- I moved the build detection results to normal preprocessor definitions (changes in setup.py and some declarations in pyfftw.pxd)
- Some of the conditions surrounding regular code can just be changed to normal conditions
- For functions that are conditionally available I added macros in pyfftw.pxd that will replace the missing symbols.

With the added macros, the Cython cdef wrapper functions can be defined unconditionally. They will just not be called at runtime since the same macros will also ensure they aren't registered (in `planners`, `destroyers`, etc.)

Then I also just replaced a few other functions that had deprecation warnings (`LooseVersion` to `packaging.Version` and `numpy.alltrue` to `numpy.all` in the tests)

JeppeKlitgaard commented:
> I've only been able to take brief look. It seems like it may have been left commented out inadvertently. That change was made as part of [bfde139](https://github.com/pyFFTW/pyFFTW/commit/bfde139db9cfbd09aaa5cb3db8dce5263f546b5d) which seems to be focused on changing how stdout/stderr are captured, and I didn't see a mention of changes to the flags argument in the associated PR.

I believe I am to blame here – I can't recall why that line was commented out. You're likely correct that this was inadvertent and an oversight on my part. Apologies about the inconvenience

karlotness commented:
No worries! It's very subtle and seems to only impact builds against static libraries which I'd imagine are much less common (and on Linux only too), and so is much harder to detect.

hgomersall commented:
Are we good to go with this once it's reviewed?

karlotness commented:
I think this should be set, yes.

The one thing that might need some separate attention for Python 3.12 might be versioneer. Looks like there are some issues with older vendored copies (conda-forge/pyfftw-feedstock#61)

karlotness commented:
I spoke a bit too soon. *Now* it should actually be ready :slightly_smiling_face: 

I just saw #364 it looks that PR adds noexcept on a number of fftw functions. I think that's a good idea and it also silences a number of warnings from Cython about reaquiring the GIL for exceptions checks. I went through and added those to all the `_fftw` functions that just call through into FFTW since none of those will set Python exceptions.

I also noticed that the `set_timelimit` and `plan_with_nthreads` functions weren't handled like the others--they should be handled the same way now.

paugier commented:
It seems to me that this PR is useful and should be reviewed!

hgomersall commented:
I'm happy with this now (having taken the opportunity to understand it!)

hgomersall commented:
I've merged this because it made sense, but it's still failing CI, which looks like it's because the cython version is demanding the latest version of Numpy which is restricted to python >3.8. What's the right thing to do here? Do we drop support for 3.8, or restrict the Cython version?

karlotness commented:
It looks like those build errors might be caused by some small conflicting changes. I've opened a new PR (#369) that should fix those.

karlotness commented on pyfftw/pyfftw.pxd:
```diff
@@ -47,6 +63,70 @@ cdef extern from 'pyfftw_complex.h':
     ctypedef long double clongdouble[2]
 
 cdef extern from 'fftw3.h':
+    """
```
That's right, the block there should insert those raw macro definitions right after including fftw3 and those should keep the build working even without the actual functions being available, but they also shouldn't be called since they won't be added to the lookup tables since that's decided by the same conditions.
